### ISSUE TITLE
resume() does not work if playback was paused for some time in lockscreen 

### ISSUE THREAD

**Issue by funkenstrahlen** [delannoyk/AudioPlayer#21]
I have discovered the following problem:
- call player.playItem() wait until it is playing

```
Stopped -> Buffering
Buffering -> Playing
```
- lock phone
- pause stream `Playing -> Paused`
- wait some time
- try to play again

```
Paused -> Playing
Playing -> Buffering
```

It will stay in buffering, but does not play audio.

I think this is a problem with `AVPlayer`, because `resume()` looks like this:

``` swift
    public func resume() {
        player?.play()
        state = .Playing
    }
```

`player.play()` is not enough to recover the stream because I think `AVPlayer` cuts the connection very fast when the phone is locked. This behaviour does not happen when the phone is not locked.

I realised calling `player.playItem()` instead of `player.resume()` does actually work from lockscreen.

I do not know what's the best way to address this issue. The player seems to "realize" that it has to start buffering as it moves into `.Buffering` immediately. However it does not handle the case that `AVPlayer` does not actually start playing.

**Comment by funkenstrahlen** [comment]
I have to update the description above. I see a similar problem when playback is paused because another app starts playing audio.

When I then come back to my app after some time and player.resume() is called it does not play any more.

**Comment by delannoyk** [comment]
Mh. I'll see when I'm done implementing #8 if it's still an issue since I'll handle the connection and not AVPlayer.

Is this a major issue for you to use AudioPlayer - I guess it's but how urgent is this? I still have some issue with #8 implementation and it's not ready right now.

**Comment by funkenstrahlen** [comment]
I do not know enough about your code to know why this is happening and what's the best approach to address the issue.

It is not very urgent though.

What are you changing because of #8 that might fix that?

**Comment by funkenstrahlen** [comment]
Any news on this?

**Comment by delannoyk** [comment]
Hi @funkenstrahlen 
Just so you know, this is my next priority right after releasing what's on `feature/event_producers` which should come soon.

**Comment by delannoyk** [comment]
Hi @funkenstrahlen 
I'm working on this now :)
Do you remember what "some time" means in this context - like 1min/10mins/1hour?
> * lock phone
> * pause stream Playing -> Paused
> * wait some time
> * try to play again

**Comment by delannoyk** [comment]
Don't bother answering, it stops buffering almost instantaneously.

**Comment by delannoyk** [comment]
From what I see on Charles Proxy, the problem is that AVPlayer cuts the connection to the live stream after some seconds when in background --> There is no background task living there because the player doesn't emit any sound anymore so the `UIBackgroundMode` `audio` doesn't apply.

Although I'd like to have the exact same behaviour as when not in background, it won't be possible for long pauses, since `beginBackgroundTaskWithExpirationHandler` only allow around ~10mins.
What I suggest instead, is to add to `resume()` a retryTimer that will of course retry playing the item (by creating a new AVPlayer etc - see `retryOrPlayNext()`) if the player doesn't start playing within the `retryTimeout` interval.

What do you think about this?

Already tested it and I can say it works, although the user experience is lowered because the play button doesn't do anything immediately - you could lower `retryTimeout` in order to make it quicker to resume.

**Comment by funkenstrahlen** [comment]
You solution sounds quite good, but I have some questions to be sure I understand correctly what you want to do:

It is perfectly ok for the player to be killed after some time when paused in background. That's what iOS does. Currently this time seems to be very short as you also discovered. Also the player controls are still visible in the lockscreen, so I assume the app is not yet killed after 1 minute. If you can get the resume working for about 10min in lockscreen that would be awesome. After that time the lockscreen controls should be gone anyway.

The other question is what happens if the user opens the app again to listen again. I think autoplay is no good idea in this case. However it is important that doing a `resume()` when the playbutton is pressed actually recovers the stream (or player goes to failed if the stream is gone).

So in the end this is somehow the same question as both actions call `resume()` right?

Using retry count is a good solution. However the player should go to `buffering`, so I can show the user something is happening (not possible in lockscreen of course). Adjustable the timeout is a good idea.

**Comment by funkenstrahlen** [comment]
just a thought as you work with many `NSTimers`: Is there any timer recovery required when the app comes back from being killed in background?

**Comment by delannoyk** [comment]
I only work with 1 timer, max 2: one for adjusting quality if necessary, another one when retrying.
> just a thought as you work with many NSTimers: Is there any timer recovery required when the app comes back from being killed in background?

Nope.

The problem is not that the app is killed or put in sleep while in background, it's that AVPlayer kills the connection with the steam on pause in background if no background task is started. As said earlier, starting a background task in `pause()` fixes the issue for the time background task is alive, not further.
Also, the problem also exists if you do this:
1. Open app and start playing
2. Wait for player to start
3. Go to lock screen & pause
4. Wait like ~20-30 seconds (I usually wait for the screen to turn off)
5. Open app, resume()

At this point, the player won't stream anything although AVPlayer will say it's buffering, it's not (and it won't `resume()` unless the user or the app developer explicitly asks to, so no worries there.

> Using retry count is a good solution. However the player should go to buffering, so I can show the user something is happening (not possible in lockscreen of course). Adjustable the timeout is a good idea.
It's already going to buffering mode, actually it does:
1. Paused (from the lock screen)
2. `resume()` (by developer or user) -> .Playing
3. Almost instantaneously -> .Buffering
4. --> AudioPlayer starts a timer to retry if the player doesn't start in the given delay.

As stated, starting a background task would also work even though I don't think this is the way to go.

Let me push this fix to develop so you can try and let me know.

**Comment by delannoyk** [comment]
Pushed on `develop`.

Note that this fix is not a hundred percent finished, I still have some more testing to do :)

**Comment by funkenstrahlen** [comment]
Great. I will have time to test it tomorrow.

**Comment by funkenstrahlen** [comment]
Left a comment on the commit 97ec768ceb842e1068945a6e15bd0034bc70d199

**Comment by delannoyk** [comment]
Pushed to develop. I added both a retry and a background task because I feel like mimicking AVPlayer's default pause behaviour is important.
I'll push it to master and create a new Pod release for it a bit later tonight.

**Comment by funkenstrahlen** [comment]
when do you plan to release the fix as a new version?

**Comment by delannoyk** [comment]
Tonight or tomorrow morning based on when I get home :) Thanks for the PR btw

**Comment by funkenstrahlen** [comment]
I did not perform any big tests yet. So you might want to try it first before you release it.

**Comment by funkenstrahlen** [comment]
I discovered a problem with this line in `resume()`:

```swift
        //We don't wan't to change the state to Playing in case it's Buffering. That
        //would be a lie.
        if state != .Playing && state != .Buffering {
            state = .Playing
        }
```

If the player is paused in background for some time its state is `.Paused`. This will reset it to `.Playing` and therefore the retryTImer will fail because it does not retry if the state is already playing.

**Comment by funkenstrahlen** [comment]
shouldn't the state only be changed by AVPlayer delegate methods?

**Comment by delannoyk** [comment]
If the state is still `.Playing` that means the `playbackBufferEmpty` is not fired. So yes, I guess I should change that.
Working on it.

**Comment by funkenstrahlen** [comment]
> I don't know why AVPlayer wouldn't fire it in case it was buffering (even though the connection has been cut off).

Couldn't you reproduce it?

**Comment by funkenstrahlen** [comment]
I will try to further debug it. Maybe I can find the cause...

**Comment by funkenstrahlen** [comment]
This looks like a problem with AVPlayer. I though about adding a KVO on `player.rate`. However this also does not help.

In `resume()` the `player.rate` ist set to `1.0` which should make AVPlayer play again. If it can not play, because it was in background for too long, it simply does not play, but `player.rate` stays at `1`. So maybe that is why it does not fire the `playbackBufferEmpty`?

**Comment by delannoyk** [comment]
Hi @funkenstrahlen 
Sorry I have very few time this week to look into it (deadline, you know :P).
Anyway, for now, I think I'm going to remove the `retry` function on resume and just let the background task so that we can release a fix for this issue and you're not blocked anymore by this bug. If I recall correctly, you were in favor of only this?

> If you can get the resume working for about 10min in lockscreen that would be awesome.

I'll readd this when I have more time to work outside... work :P

**Comment by funkenstrahlen** [comment]
I really understand you can not work on this right now. I have no deadline, so no problem.

I do not think removing `retry` is a good idea. Without it the current `resume` does not even work for 1 minute in background. The app keeps running fine, but the AVPlayer has problems to recover.

**Comment by delannoyk** [comment]
Thank you, once again, for your contribution on this project. It's really appreciated.

**Comment by funkenstrahlen** [comment]
I found these problem reports:

* http://stackoverflow.com/questions/36617232/avplayer-cant-resume-after-paused-some-waiting
* http://stackoverflow.com/questions/36858988/avplayer-does-not-fire-playbackbufferempty-but-does-not-play-either (opened this myself)
* http://stackoverflow.com/questions/36781386/avplayer-pause-live-stream-on-ios-9-3

**Comment by funkenstrahlen** [comment]
There is also another problem I see. This is only relevant if AVPlayer worked like expected! 

* set player.rate to 1
* state to playing
* retryTimer called. does not do anything as it sees state = playing
* playBackBufferEmpty called. settings state to buffering

This happens if the `retryTimeout` is very short.

**Comment by funkenstrahlen** [comment]
I build a Playground to reproduce it. You can find the code here if you want to experiment: http://stackoverflow.com/questions/36858988/avplayer-does-not-fire-playbackbufferempty-but-does-not-play-either

**Comment by funkenstrahlen** [comment]
I see a big problem with `AVPlayer` which makes it very hard to fix this.

I can see this variable configuration in two cases: the player is playing audio and it is not.
```
rate: 1.0
buffer empty? true
buffer full? false
playback keeping up? true
item status: READY
player status: READY
state: PLAYING
```

This makes it **impossible** to actually know if the player plays audio.

**Comment by funkenstrahlen** [comment]
I found a solution but it has some compromise:

Retry until `playback keeping up = true` and the buffer it not empty.

The little problem: If the playback managed to resume before with a buffer reported as empty the retry will still reset the item until the buffer manages to fill up. this adds little audio breaks.

**Comment by funkenstrahlen** [comment]
@delannoyk If you like you can take a look at this: https://github.com/funkenstrahlen/AudioPlayer/commit/01169ee5e886e46981e0fbe15800cb804234be6e

That's what I tried. But it seems impossible to make it work as long as AVPlayer does not report any indication if its playing or not.

**Comment by funkenstrahlen** [comment]
I filed a bug report.

**Comment by funkenstrahlen** [comment]
still no reaction to my radar bug report from apple.

**Comment by delannoyk** [comment]
Hi,
Thanks for all the information you gave and actions you took.
Any chance you have an openradar issue number so that I can dupe?

**Comment by funkenstrahlen** [comment]
@delannoyk I published my radar to openradar: http://openradar.appspot.com/radar?id=4970939146043392

**Comment by funkenstrahlen** [comment]
@delannoyk Did you also file a radar? Did you get feedback from Apple?

**Comment by delannoyk** [comment]
@funkenstrahlen Yes I did and no feedback at all 😞

**Comment by delannoyk** [comment]
A colleague of mine is going to WWDC with the number of this radar 😄

**Comment by funkenstrahlen** [comment]
I just watched the AVFoundation Session from WWDC. Looks like they fixed a lot of this in AVPlayer.

**Comment by delannoyk** [comment]
Hey @funkenstrahlen 
I got an update from Apple saying it got fixed in 11.0.2 and the playground seems to be working now, haven't tried on device yet.
Cheers
